lib/bolt/module_installer/specs.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+require_relative '../error'
+require_relative 'specs/forge_spec'
+require_relative 'specs/git_spec'
+
+module Bolt
+  class ModuleInstaller
+    class Specs
+      def initialize(specs = [])
+        @specs = []
+        add_specs(specs)
+        assert_unique_names
+      end
+
+      # Creates a list of specs from the modules in a Puppetfile object.
+      #
+      def self.from_puppetfile(puppetfile)
+        new(puppetfile.modules.map(&:to_hash))
+      end
+
+      # Returns a list of specs.
+      #
+      def specs
+        @specs.uniq(&:name)
+      end
+
+      # Returns true if the specs includes the given name.
+      #
+      def include?(name)
+        _owner, name = name.tr('-', '/').split('/', 2)
+        @specs.any? { |spec| spec.name == name }
+      end
+
+      # Adds a spec.
+      #
+      def add_specs(*specs)
+        specs.flatten.map do |spec|
+          case spec
+          when Hash
+            @specs.unshift spec_from_hash(spec)
+          else
+            @specs.unshift spec
+          end
+        end
+      end
+
+      # Parses a spec hash into a spec object.
+      #
+      private def spec_from_hash(hash)
+        return ForgeSpec.from_hash(hash) if ForgeSpec.implements?(hash)
```

Comment:
It seems like using a hash as an intermediary representation of this data might be unnecessary - translating objects from the resolver directly into our own objects would probably be best.

Reply:
This is for the specs from project config, which are already hashes and don't come from the resolver.